SWIPs/swip-adaptive-node-protocol.md

```diff
@@ -0,0 +1,119 @@
+## Abstract
+
+Adaptive node operation means a node should be able to change the capabilities it has according to changes in available resources.
+
+## Motivation
+
+The peers of the node may need to adjust their behavior towards the node when these capabilities change. Therefore, a protocol is required to inform the peers of changes in capabilities.
+
+The motivation is critical for SWIPs that want to change the Swarm protocol. It should clearly explain why the existing protocol specification is inadequate to address the problem that the SWIP solves. SWIP submissions without sufficient motivation may be rejected outright.
+
+## Specification
+
+Upon initial connection, capabilities can be transmitted already within the handshake message.
+
+Capabilities are combinational. Some combination may make more sense than other. Interpreting which combinations are feasible and which are not should be handled by respective nodes, and not be hardwired in the protocol.
+
+For this reason, a bit vector seems the most logical choice. The bit vector can represent the adaptive node operation state within the node, and be embedded into the protocol message as-is.
+Since the capabilities may have to be set from different system modules, the network module should provide an API for registering them. The following structure is proposed:
+
+* the first byte is the identifier of the module (0x00 is bzz, and built in)
+* the second byte is the length `n` of the bit vector containing the flags
+* the last `n` bytes is the bit vector containing the flags
+
+For the individually capability flags we will initially be choosing from, we can see an example of two distinct byte pairs representing tematic categories:
+
+```
+BZZ/SYNC: 			// (registered by default)
+- retrieve		0x00020001
+- push			0x00020002
+- retrieve relay	0x00020010
+- push sync relay	0x00020020
+- storer node		0x00028000	// the node participates as storer for area of responsibility
+
+PSS:
+- receive	0x010101
+- send		0x010102
+- relay		0x0l0110
+```
+
+We define a new protocol message for the `bzz` protocol called `Capabilities`. This message contains one field holding all module capability bit vectors.
+
+```
+Capabilities {
+	Module [][]byte
+}
+```
+
+This protocol message can be sent at any time to inform peers of _changes_ to the node's capabilities. 
+
+Furthermore, the bzz handshake message itself should also embed this message instead of the `Light` field that is currently there. The `HandshakeMsg` message this becomes:
+
+```
+HandshakeMsg {
+	Version		uint64
+	NetworkID	uint64
+	Addr		*BzzAddr
+	Capabilities	Capabilities
+}
+```
+
+For simplicity, arrays of bools are used to represent the flags in interfacing code. In the JSON-RPC API a `Capability` is represented in the following manner:
+
+```
+{
+	"Id": <int>,
+	"Cap": []<bool>,
+}
+```
+
+For example, a `Capability` with id 42 and three flags with values `true, false, true` this becomes:
+
+```
+{
+	"Id": 42,
+	"Cap": [true, false, true],
```

Comment:
How extensible is this if capabilities are only encoded as bools in a list? What about using enums and something like protobuf? That way they capabilities can easily be added and dealt with as optional args

Reply:
Yes they were initially just bit vectors, but changed due to PR comments on the golang implemenation. But using bools is far from optimal also for performance reasons, so I will change this back.